v2/event/event_unmarshal.go

```diff
@@ -0,0 +1,447 @@
+						cachedData = stream.Buffer()
+						err = stream.Error
+						appendFlag(&state, dataBase64Flag)
+					default:
+						value := val.GetInterface()
+						if ctx.Extensions == nil {
+							ctx.Extensions = make(map[string]interface{}, 1)
+						}
+						err = ctx.SetExtension(tokenQueue.slice[i].key, value)
+					}
+					if err != nil {
+						return err
+					}
+				}
+			}
+			continue
+		}
+
+		// If no specversion, enqueue unconditionally
+		if !checkFlag(state, specVersionV03Flag|specVersionV1Flag) {
+			tokenQueue.push(key, iterator.ReadAny())
+			continue
+		}
+
+		// From this point downward -> we can assume the event has a context pointer non nil
+
+		// If it's a datacontenttype, trigger state change
+		if key == "datacontenttype" {
+			if checkFlag(state, dataContentTypeFlag) {
+				return fmt.Errorf("datacontenttype was already provided")
+			}
+
+			dct := iterator.ReadString()
+
+			switch ctx := out.Context.(type) {
+			case *EventContextV03:
+				ctx.DataContentType = &dct
+			case *EventContextV1:
+				ctx.DataContentType = &dct
+			}
+			appendFlag(&state, dataContentTypeFlag)
+			continue
+		}
+
+		// If it's a datacontentencoding and it's v0.3, trigger state change
+		if checkFlag(state, specVersionV03Flag) && key == "datacontentencoding" {
+			if checkFlag(state, dataBase64Flag) {
+				return ValidationError{"datacontentencoding": errors.New("datacontentencoding was specified twice")}
+			}
+
+			dce := iterator.ReadString()
+
+			if dce != Base64 {
+				return ValidationError{"datacontentencoding": errors.New("invalid datacontentencoding value, the only allowed value is 'base64'")}
+			}
+
+			out.Context.(*EventContextV03).DataContentEncoding = &dce
+			appendFlag(&state, dataBase64Flag)
+			continue
+		}
+
+		// We can parse all attributes, except data.
+		// If it's data or data_base64 and we don't have the attributes to process it, then we cache it
+		// The expanded form of this condition is:
+		// (checkFlag(state, specVersionV1Flag) && !checkFlag(state, dataContentTypeFlag) && (key == "data" || key == "data_base64")) ||
+		// (checkFlag(state, specVersionV03Flag) && !(checkFlag(state, dataContentTypeFlag) && checkFlag(state, dataBase64Flag)) && key == "data")
+		if (state&(specVersionV1Flag|dataContentTypeFlag) == specVersionV1Flag && (key == "data" || key == "data_base64")) ||
+			((state&specVersionV03Flag == specVersionV03Flag) && (state&(dataContentTypeFlag|dataBase64Flag) != (dataContentTypeFlag | dataBase64Flag)) && key == "data") {
+			if key == "data_base64" {
+				appendFlag(&state, dataBase64Flag)
+			}
+			cachedData = iterator.SkipAndReturnBytes()
+			continue
+		}
+
+		// At this point or this value is an attribute (excluding datacontenttype and datacontentencoding), or this value is data and this condition is valid:
+		// (specVersionV1Flag & dataContentTypeFlag) || (specVersionV03Flag & dataContentTypeFlag & dataBase64Flag)
+		switch ctx := out.Context.(type) {
+		case *EventContextV03:
```

Comment:
we could flatten this out a bit and have only the fields that change for versions inside the case statement like:

```
switch key {
  case "id":
    <do it>
  case :type":
    <do it>
...
  default:
  switch ctx := out.Context.(type) {
		case *EventContextV03:
                    switch key {
                       "schemaurl":
                    }
... ect ?
}
```

Reply:
Actually, There does not seem to be a difference between the common keys and the uncommon keys 


Ohg wait. no sorry... ctx is typed. NEVERMIND.

Reply:
I think there are getter/setter methods on `EventContextV*` so maybe this idea is back on the table?

Reply:
These setters are quite costly, i would rather prefer to avoid them here and perform direct access to the context. We can improve those setters and revisit this code later, is that ok for you?